RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,304 @@
+# 0031 Multi Test Run Finalization
+
+# Summary
+This document details a data collector extensibility point to reporcess (combine/merge) attachments obtained across test executions in both design mode and commandline scenarios.
+
+# Motivation
+Today when Test Platform executes tests in parallel only code coverage reports are merged (data collector attachments with uri: `datacollector://microsoft/CodeCoverage/2.0`). For other data collector attachments reprocessing is skipped and all of them are returned by Test Platform.
+
+The [dotnet test](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test) command is used to execute unit tests in a given solution. The `dotnet test` command builds the solution and runs a test host application for each test project in the solution. However, currently there is no way to reporcess (combine/merge) data collector attachments associated with each project execution. Code coverage reports are not merged.
+
+When `Run All Tests` is performed in VS, tests for projects can be executed separately based on target platform/target framework amongst other criteria. In this case also combining/merging of data collector attachments is not performed (e.g. code overage reports are not merged). `Analyze Code Coverage for All Tests` is showing coverage report for only some test projects in the run.
+
+# Proposed Changes
+
+Introduce a new `IDataCollectorAttachmentProcessor` interface which can be implemented by Test Platform extensions and provide custom logic to reporcess (combine/merge) data collector attachments. Below interface should be used only for providing logic to reprocess information from data collector attachments from independent test executions. In other words data collector attachment should be modified/created using data from other data collector attachments. This interface should **not be** used for modifying any single data collector attachment. Test Platform will invoke `ProcessAttachmentSets` only if at least 2 data collector attachments related to processor (through `GetExtensionUris`) are created by independent test executions.
```

Comment:
>IDataCollectorAttachmentProcessor [](start = 17, length = 33)

We'd also want to add how these processors would be discovered. Would it be through runsettings? Would they need to be copied over to the output directory? Would they need to be placed in the Extensions folder? I'd prefer if we do not require these extensions to ship with TestPlatform.

Reply:
@AbhitejJohn  I don't have any experience with this. What do you recommend?

Reply:
It wouldn't be unreasonable, for this effort, to make this similar to our data collector extension discovery - which allows extension discovery through a testadapterpath specification in runsettings/commandline or through the extensions folder. As part of the performance efforts, we have another work item to optimize sending in the relevant extensions to the platform which should cover better nuget package support as well. 

---
In reply to: [447196449](https://github.com/microsoft/vstest-docs/pull/235#discussion_r447196449) [](ancestors = 447196449)

Reply:
@AbhitejJohn below my thinking, I can be totally wrong :)
1) For IDE before running finalization we know all test requests, so we know runsettings for projects. We will take all testadapter paths from all runsettings and we can call InitializeExtensions before running finalization
2) For commandline we will somehow figure out extensions paths (by getting through commandline or finding all runsettings for all projects). Can those paths be sent together with attachments and test statistics? 
When we have all paths we can always before running finalization call InitializeExtensions.

Another possibility is to add extension paths as parameter to FinalizeMultiTestRunAsync on VsTestConsoleWrapper.

Reply:
So for IDE we already call InitializeExtensions based on the test adapters we find in the solution before a test discovery/execution. I'd recommend we reuse those. I'm assuming also that we are changing the finalization request to take in runsettings in which case the test platforms adapter search logic can be used to process that as well. 

The test platform also looks up extensions from the output directory if they aren't specified. I'd recommend using a file name pattern like we have for our other extensions *testadapter*.dll/*datacollector*.dll/*logger*.dll etc to easily find interesting extension assemblies. Perhaps *attachmentprocessor*.dll would work.

Reply:
I see in IVstestConsoleWrapper we have `string runSettings` for test runs and `string discoverySettings` for discovery. In case of finalization which runsettings we will use if we can several test projects with different runsettings. Maybe we could add parameter `IEnumerable<string> pathToAdditionalExtensions` to Finalize method?

Reply:
added processingSettings